invenio_banners/services/service.py

```diff
@@ -7,28 +7,61 @@
 
 """Banner Service API."""
 
-from invenio_access.permissions import system_identity
 from invenio_records_resources.services import RecordService
+from invenio_records_resources.services.base import LinksTemplate
 
-from invenio_banners.records.models import BannerModel
+from ..records.models import BannerModel
+from ..services.errors import BannerNotExistsError
 
 
 class BannerService(RecordService):
     """Banner Service."""
 
-    def read(self, url_path):
+    def read(self, identity, id):
         """Retrieve a banner."""
-        # TODO: not implemented/tested yet
-        # self.require_permission(identity, "read")
-        banner = BannerModel.get_active(url_path)
-        return self.result_item(self, system_identity, banner)
-
-    def search(self, identity, params=None, **kwargs):
-        """Search for resources matching the querystring."""
-        # TODO: not implemented/tested yet
-        # self.require_permission(identity, "read")
-        banners = BannerModel.search()
-        return self.result_list(self, system_identity, banners)
+        # resolve and require permission
+        self.require_permission(identity, "read")
+
+        banner = self.record_cls.get(id)
+
+        # check if banner exists
+        if banner is None:
+            raise BannerNotExistsError(id)
+
+        return self.result_item(
+            self,
+            identity,
+            banner,
+            links_tpl=self.links_item_tpl,
+        )
+
+    def read_active(self, identity, active, url_path, params=None):
+        """Retrieve a banner."""
+        # resolve and require permission
+        self.require_permission(identity, "read")
+
+        banners = self.record_cls.get_active(active, url_path)
+
+        return self.result_list(
+            self,
+            identity,
+            banners,
+            links_tpl=LinksTemplate(self.config.links_search, context={"args": params}),
+            links_item_tpl=self.links_item_tpl,
+        )
+
+    def search(self, identity, params=None):
+        """Search for banners matching the querystring."""
+        self.require_permission(identity, "search")
+
+        banners = BannerModel.query.all()
```

Comment:
Warning: if there is a high number of banners, this can be slow.
In general, we use a search engine to avoid a series of issues related to retrieving a list of results from the DB.
We should discuss this IRL.